drivers: led_strip: Add UART-based driver for WS2812

waihongtam opened the pull request:
This commit introduces a new driver for WS2812 and compatible LED strips that uses a UART peripheral.

The driver generates the precise, high-speed signal required by the WS2812 protocol by encoding each data bit into a multi-bit "symbol" and using a frame-aware packing strategy for transmission:
 - Signal Inversion: The UART's TX line must be inverted (tx-invert) to create the protocol's required idle-low signal. A UART start bit then generates the initial high pulse of a WS2812 bit.
 - Frame-Aware Packing: The driver reuses the UART's hardware- generated start and stop bits as the first and last bits of the on-wire symbol. The inner bits of the symbol are packed into the UART data payload.

This packing scheme imposes a configuration constraint: the total number of bits in a UART frame (1 start + N data + 1 stop) must be an integer multiple of the symbol's length (bits-per-symbol).

waihongtam committed:
drivers: led_strip: Add UART-based driver for WS2812

This commit introduces a new driver for WS2812 and compatible LED
strips that uses a UART peripheral.

The driver generates the precise, high-speed signal required by the
WS2812 protocol by encoding each data bit into a multi-bit "symbol"
and using a frame-aware packing strategy for transmission:
 - Signal Inversion: The UART's TX line must be inverted (tx-invert)
   to create the protocol's required idle-low signal. A UART start
   bit then generates the initial high pulse of a WS2812 bit.
 - Frame-Aware Packing: The driver reuses the UART's hardware-
   generated start and stop bits as the first and last bits of the
   on-wire symbol. The inner bits of the symbol are packed into the
   UART data payload.

This packing scheme imposes a configuration constraint: the total
number of bits in a UART frame (1 start + N data + 1 stop) must be
an integer multiple of the symbol's length (bits-per-symbol).

Signed-off-by: Wai-Hong Tam <waihong@google.com>

dsseng commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
Could you please commit dt bindings separately from (and before) the driver?

fabiobaltieri commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,342 @@
+	struct k_sem tx_done_sem;
+};
+
+static inline const struct ws2812_uart_cfg *dev_cfg(const struct device *dev)
+{
+	return dev->config;
+}
+
+static inline struct ws2812_uart_data *dev_data(const struct device *dev)
+{
+	return dev->data;
+}
```
drop, just use dev->data and dev->config directly, this does not really make the code any more compact (you don't even need a casting) and it's not normally done for Zephyr drivers

waihongtam committed:
dts: bindings: led: Add worldsemi,ws2812-uart binding

Add a new devicetree binding, worldsemi,ws2812-uart, for controlling
WS2812-compatible LED strips using a UART peripheral.

Signed-off-by: Wai-Hong Tam <waihong@google.com>

fabiobaltieri commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,342 @@
+	struct ws2812_uart_data *data = dev_data(dev);
+	const size_t total_symbol_bits =
+		num_pixels * cfg->num_colors * BITS_PER_COLOR_CHANNEL * cfg->bits_per_symbol;
+	const size_t buf_len = DIV_ROUND_UP(total_symbol_bits, cfg->uart_frame_bits);
```
^^ still relevant

waihongtam replied:
[Opps, I accidentally deleted my previous comment; I tried to summarize it shortly to keep a record]

The `led_strip.h` API ensures `num_pixels` is less than or equal to `chain-length`, which implies that `buf_len` is also less than or equal to `WS2812_UART_BUFSZ` because both of these two sizes are calculated in the same way (by multiplying and dividing by the same constants). An extra size check is not necessary.

I improved the calculation using a macro, which should make the above implication clearer.

thedjnK commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,75 @@
+      The number of UART bits used to represent a single WS2812 data bit.
+      The value must be between 3 and 10, inclusive.
+
+  uart:
```
this should be a child node of the UART rather than having a reference to a UART, e.g. an i2c sensor is a child of an i2c controller, it's not a separate entry that has a phandle reference to the i2c controller

waihongtam replied:
It is a good idea. Fixed it like this.

thedjnK commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -28,6 +28,17 @@ config WS2812_STRIP_SPI_FORCE_NOCACHE
+	select UART
+	select UART_ASYNC_API
+	help
+	  Enable driver for WS2812 (and compatibles) LED strip using UART.
```
```suggestion
	  Enable driver for WS2812 (and compatible) LED strips using UART.
```

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,336 @@
+ * packing strategy to transmit these symbols efficiently.
+ *
+ * Signal Inversion:
+ * The WS2812 protocol requires an idle-low signal. This is achieved by
+ * inverting the UART's TX output (requiring the "tx-invert" devicetree
+ * property).
```
would probably be good to check the parent device to see if this flag is set? Though not sure if it's a general UART flag or different for each UART

waihongtam replied:
The check is added below:

```
#define DT_UART_HAS_TX_INVERT(inst) (DT_PROP_OR(DT_UART_NODE(inst), tx_invert, 0) != 0)
```

```
        BUILD_ASSERT(DT_UART_HAS_TX_INVERT(idx),                                                   \
                     "The UART peripheral must be configured with tx-invert.");                    \
```

It is a general UART flag but optional.

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,336 @@
+/*
+ * Helper macros to get UART frame configuration from the referenced UART's devicetree node.
+ */
+#define _DT_UART_NODE(inst)          DT_INST_PHANDLE(inst, uart)
+#define _DT_UART_DATA_BITS(inst)     DT_PROP_OR(_DT_UART_NODE(inst), data_bits, 8)
+/* Only UART_CFG_STOP_BITS_1 or 1 is supported. Other values will fail the BUILD_ASSERT below. */
+#define _DT_UART_STOP_BITS(inst)     DT_ENUM_IDX_OR(_DT_UART_NODE(inst), stop_bits, 1)
+#define _DT_UART_HAS_PARITY(inst)    (DT_ENUM_IDX(_DT_UART_NODE(inst), parity) != \
+				      UART_CFG_PARITY_NONE)
+#define _DT_UART_HAS_TX_INVERT(inst) (DT_PROP_OR(_DT_UART_NODE(inst), tx_invert, 0) != 0)
```
these are local to this file to don't need a _ prefix

waihongtam replied:
Fixed them.

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,336 @@
+	uint8_t frame_bit_pos = 0;
+	int ret;
+
+	/* Lock the driver to ensure thread-safe access to the buffer and UART */
+	k_mutex_lock(&data->lock, K_FOREVER);
```
can this be a semaphore?

fabiobaltieri replied:
Sempahores are normally preferred for driver code, few reasons for that, Andy goes in more details in https://youtu.be/p8OgqVQxklo?si=iaVjKvUwvntmViGn&t=1304 but yeah if you don't need priority inversion semaphore seems to be the way to go, also since most core code do not use mutexes the code has a chance of being compiled out if we keep not using it in drivers.

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,336 @@
+
+	for (int i = 0; i < cfg->num_colors; i++) {
+		switch (cfg->color_mapping[i]) {
+		case LED_COLOR_ID_WHITE:
```
if not supported as above, then should return an error?

simonguinot reviewed (commented):
Hi @waihongtam,

Thanks for this very nice driver. I don't have much to say. Please see my comments below.

Additionally, I'll test it on an Everlight B1414 controller. The timings are different from the Worldsemi WS2812 (4 bits are needed to represent a "zero" or "one" bit).

waihongtam committed:
tests: drivers: build_all: led_strip: Add build test for ws2812-uart driver

This commit adds a build test for the new `worldsemi,ws2812-uart`
driver. A devicetree node for the driver is added to the app.overlay,
instantiated on a `vnd,serial` test UART.

Signed-off-by: Wai-Hong Tam <waihong@google.com>

waihongtam committed:
dts: bindings: test: Add pin inversion to vnd,serial binding

Add the `uart-controller-pin-inversion.yaml` include to the `vnd,serial`
test binding. This enables testing for other drivers that require pin
inversion to function correctly. For example, the
`worldsemi,ws2812-uart` LED strip driver relies on `tx-invert`.

Signed-off-by: Wai-Hong Tam <waihong@google.com>

waihongtam commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,66 @@
+
+  The internal memory buffer size is determined by `bits-per-symbol`. For
+  each bit of pixel data, the driver generates a symbol of N bits (where N
+  is the value of `bits-per-symbol`), which are then packed into the buffer.
```
This is a good suggestion. Added this to the description.

waihongtam commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,336 @@
+/* Calculate the buffer size needed. */
+#define WS2812_UART_CALC_BUFSZ(num_px, num_colors, bits_symbol, bits_frame)                        \
+	DIV_ROUND_UP((num_px) * (num_colors) * BITS_PER_COLOR_CHANNEL * (bits_symbol),             \
+		     (bits_frame))
```
Your analysis is correct. The macro calculates the number of UART frames, which is then used for the buffer size. This works because of the 1:1 mapping between bytes in the `uart_tx()` buffer and frames transmitted by the UART hardware. For each `uint8_t` provided to the API, the peripheral sends one complete frame (`start` + `data` + `stop`).

So, the total number of frames is equal to the number of bytes we allocate for the buffer.

waihongtam replied:
You are correct. When the UART is configured with data-bits less than 8 (e.g., 7), the MSB(s) of each byte in the tx buffer are unused. This is a direct consequence of the `uart_tx()` API, which is designed to process a buffer where each byte corresponds to a single UART frame's data payload. The hardware then transmits only the configured number of LSBs from each byte (e.g., 7 out of 8 bits).

simonguinot replied:
Thanks !

waihongtam commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,66 @@
+  The internal memory buffer size is determined by `bits-per-symbol`. For
+  each bit of pixel data, the driver generates a symbol of N bits (where N
+  is the value of `bits-per-symbol`), which are then packed into the buffer.
+
```
An example was added here and also as a node in the `tests/drivers/build_all/led_strip/app.overlay`.

matt-wood-ct commented:
> Hi @matt-wood-ct. Maybe you would be interested in testing this driver ?

Sorry I don't have HW to test this on at the moment, but I had a quick review of the code and it looks good to me

simonguinot reviewed (commented):
Hi @waihongtam,

Sorry for the delay. It took me a while to test the driver. I had trouble finding a board with a serial controller capable of delivering a 2.5-3 Mbps rate. And I also had trouble configuring it...

So I was able to test the driver with a strip of 18 Everlight B1414 controllers (with slightly different timings than the WS2812) connected to an LPC11U68 LPCXpresso board (with modified serial and clock drivers).

The driver looks very good and I only have one comment / question left.

simonguinot reviewed (approved):
Thanks for this nice driver. Well done !
